File: typescript/src/services/campaign-service.ts
import type pg from "pg";
import type { Campaign } from "../data/models.js";
import { CampaignState } from "../data/enums.js";
import { BrandRepository } from "../repositories/brand-repository.js";
import { CampaignRepository } from "../repositories/campaign-repository.js";
import { withTransaction } from "../db/pool.js";

export class CampaignService {
  constructor(private pool: pg.Pool) {}

  async getCampaigns(): Promise<Campaign[]> {
    const repo = new CampaignRepository(this.pool);
    return repo.findAll();
  }

  async getCampaign(id: number): Promise<Campaign | null> {
    const repo = new CampaignRepository(this.pool);
    return repo.findById(id);
  }

  async createCampaign(
    brandId: number,
    title: string,
    description: string | null,
    payoutCents: number,
    maxSubmissions: number,
  ): Promise<Campaign> {
    return withTransaction(this.pool, async (client) => {
      const brandRepo = new BrandRepository(client);
      const brand = await brandRepo.findById(brandId);
      if (!brand) {
        throw new Error(`Brand ${brandId} not found`);
      }
      const campaignRepo = new CampaignRepository(client);
      return campaignRepo.create(brandId, title, description, payoutCents, maxSubmissions);
    });
  }

  async activateCampaign(id: number): Promise<Campaign> {
    return withTransaction(this.pool, async (client) => {
      const repo = new CampaignRepository(client);
      const campaign = await repo.findById(id);
      if (!campaign) {
        throw new Error(`Campaign ${id} not found`);
      }
      if (campaign.state !== CampaignState.DRAFT) {
        throw new Error("Campaign must be in DRAFT state to activate");
      }
      return repo.updateState(id, CampaignState.ACTIVE);
    });
  }

  async completeCampaign(id: number): Promise<Campaign> {
    return withTransaction(this.pool, async (client) => {
      const repo = new CampaignRepository(client);
      const campaign = await repo.findById(id);
      if (!campaign) {
        throw new Error(`Campaign ${id} not found`);
      }
      if (campaign.state !== CampaignState.DRAFT) {
        throw new Error("Campaign must be in ACTIVE state to complete");
      }
      return repo.updateState(id, CampaignState.COMPLETED);
    });
  }
}

File: python/src/services/campaign_service.py
import asyncpg

from src.data.enums import CampaignState
from src.data.models import Campaign
from src.db.pool import with_transaction
from src.repositories.brand_repository import BrandRepository
from src.repositories.campaign_repository import CampaignRepository


class CampaignService:
    def __init__(self, pool: asyncpg.Pool):
        self.pool = pool

    async def get_campaigns(self) -> list[Campaign]:
        return await CampaignRepository(self.pool).find_all()

    async def get_campaign(self, id: int) -> Campaign | None:
        return await CampaignRepository(self.pool).find_by_id(id)

    async def create_campaign(
        self,
        brand_id: int,
        title: str,
        description: str | None,
        payout_cents: int,
        max_submissions: int,
    ) -> Campaign:
        async def txn(conn: asyncpg.Connection) -> Campaign:
            brand = await BrandRepository(conn).find_by_id(brand_id)
            if brand is None:
                raise ValueError(f"Brand {brand_id} not found")
            return await CampaignRepository(conn).create(
                brand_id, title, description, payout_cents, max_submissions
            )

        return await with_transaction(self.pool, txn)

    async def activate_campaign(self, id: int) -> Campaign:
        async def txn(conn: asyncpg.Connection) -> Campaign:
            repo = CampaignRepository(conn)
            campaign = await repo.find_by_id(id)
            if campaign is None:
                raise ValueError(f"Campaign {id} not found")
            if campaign.state is not CampaignState.DRAFT:
                raise ValueError("Campaign must be in DRAFT state to activate")
            return await repo.update_state(id, CampaignState.ACTIVE)

        return await with_transaction(self.pool, txn)
